onnxruntime/core/session/inference_session.cc

```diff
@@ -172,18 +168,44 @@ void InferenceSession::ConstructorCommon(const SessionOptions& session_options,
   InitLogger(logging_manager_);  // this sets session_logger_ so that it can be used for logging after this point.
 
   // Update the number of steps for the graph transformer manager using the "finalized" session options
-  graph_transformation_mgr_.SetSteps(session_options_.max_num_graph_transformation_steps);
+  ORT_ENFORCE(graph_transformation_mgr_.SetSteps(session_options_.max_num_graph_transformation_steps).IsOK());
   use_per_session_threads_ = session_options.use_per_session_threads;
 
   if (use_per_session_threads_) {
     LOGS(*session_logger_, INFO) << "Creating and using per session threadpools since use_per_session_threads_ is true";
-    thread_pool_ = concurrency::CreateThreadPool("intra_op_thread_pool",
-                                                 session_options_.intra_op_num_threads);
-
-    inter_op_thread_pool_ = session_options_.execution_mode == ExecutionMode::ORT_PARALLEL
-                                ? concurrency::CreateThreadPool("inter_op_thread_pool",
-                                                                session_options_.inter_op_num_threads)
-                                : nullptr;
+    {
+      OrtThreadPoolParams to = session_options_.intra_op_param;
+      if (to.name == nullptr) {
+        to.name = ORT_TSTR("intra-op");
+      }
+      // If the thread pool can use all the processors, then
+      // we set thread affinity.
+      if (to.thread_pool_size == 0 && session_options_.execution_mode == ExecutionMode::ORT_SEQUENTIAL)
+        to.auto_set_affinity = true;
+      else
+        to.auto_set_affinity = false;
+      if (to.name == nullptr)
+        to.name = ORT_TSTR("intra-op");
+      thread_pool_ =
+          concurrency::CreateThreadPool(&Env::Default(), to, nullptr);
+    }
+    if (session_options_.execution_mode == ExecutionMode::ORT_PARALLEL) {
+      OrtThreadPoolParams to = session_options_.inter_op_param;
+      // If the thread pool can use all the processors, then
+      // we set thread affinity.
+      if (to.thread_pool_size == 0 && session_options_.execution_mode == ExecutionMode::ORT_SEQUENTIAL)
+        to.auto_set_affinity = true;
+      else
+        to.auto_set_affinity = false;
+      if (to.name == nullptr)
+        to.name = ORT_TSTR("intra-op");
+      inter_op_thread_pool_ =
+          concurrency::CreateThreadPool(&Env::Default(), to, nullptr);
+      if (inter_op_thread_pool_ == nullptr) {
+        LOGS(*session_logger_, INFO) << "Disabling Parallel Executor";
```

Comment:
Would be good for the user to know why you disabled even though the user requested for parallel.